softlayer/resource_softlayer_file_storage.go

```diff
@@ -0,0 +1,941 @@
+package softlayer
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/softlayer/softlayer-go/datatypes"
+	"github.com/softlayer/softlayer-go/filter"
+	"github.com/softlayer/softlayer-go/helpers/location"
+	"github.com/softlayer/softlayer-go/helpers/product"
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/session"
+	"github.com/softlayer/softlayer-go/sl"
+	"regexp"
+	"strings"
+	"time"
+)
+
+const (
+	StoragePerformancePackageType = "ADDITIONAL_SERVICES_PERFORMANCE_STORAGE"
+	StorageEndurancePackageType   = "ADDITIONAL_SERVICES_ENTERPRISE_STORAGE"
+	storageMask                   = "id,billingItem.orderItem.order.id"
+	storageDetailMask             = "id,capacityGb,iops,storageType,username,serviceResourceBackendIpAddress,properties[type]" +
+		",serviceResourceName,allowedIpAddresses,allowedSubnets,allowedVirtualGuests,snapshotCapacityGb,osType"
+	itemMask        = "id,capacity,description,units,keyName,prices[id,categories[id,name,categoryCode],capacityRestrictionMinimum,capacityRestrictionMaximum,locationGroupId]"
+	EnduranceType   = "Endurance"
+	PerformanceType = "Performance"
+	FileStorage     = "FILE_STORAGE"
+	BlockStorage    = "BLOCK_STORAGE"
+)
+
+var (
+	enduranceIopsMap = map[float64]string{
+		0.25: "LOW_INTENSITY_TIER",
+		2:    "READHEAVY_TIER",
+		4:    "WRITEHEAVY_TIER",
+		10:   "10_IOPS_PER_GB",
+	}
+
+	enduranceCapacityRestrictionMap = map[float64]int{
+		0.25: 100,
+		2:    200,
+		4:    300,
+		10:   1000,
+	}
+
+	storagePackageMap = map[string](map[string](map[string]string)){
+		FileStorage: {
+			PerformanceType: {
+				"storagePackageType":          StoragePerformancePackageType,
+				"iopsCategoryCode":            "performance_storage_iops",
+				"storageProtocolCategoryCode": "performance_storage_nfs",
+			},
+			EnduranceType: {
+				"storagePackageType":          StorageEndurancePackageType,
+				"iopsCategoryCode":            "storage_tier_level",
+				"storageProtocolCategoryCode": "storage_file",
+			},
+		},
+		BlockStorage: {
```

Comment:
`buildStorageProductOrderContainer` function supports both block_storage and file_storage.